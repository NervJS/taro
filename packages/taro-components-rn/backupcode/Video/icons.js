import * as React from 'react'
import { ActivityIndicator } from 'react-native'
import { Entypo, Foundation, MaterialIcons } from '@expo/vector-icons'

const ICON_COLOR = '#FFFFFF'
const CENTER_ICON_SIZE = 36
const BOTTOM_BAR_ICON_SIZE = 30

export const PlayIcon = () =>
  <Entypo
    name={'controller-play'}
    size={BOTTOM_BAR_ICON_SIZE}
    color={ICON_COLOR}
    style={{ textAlign: 'center' }}
  />

export const PauseIcon = () =>
  <Entypo
    name={'controller-paus'}
    size={BOTTOM_BAR_ICON_SIZE}
    color={ICON_COLOR}
    style={{ textAlign: 'center' }}
  />

export const Spinner = () =>
  <ActivityIndicator color={ICON_COLOR} size={'large'} />

export const FullscreenEnterIcon = () =>
  <MaterialIcons
    name={'fullscreen'}
    size={BOTTOM_BAR_ICON_SIZE}
    color={ICON_COLOR}
    style={{ textAlign: 'center' }}
  />

export const FullscreenExitIcon = () =>
  <MaterialIcons
    name={'fullscreen-exit'}
    size={BOTTOM_BAR_ICON_SIZE}
    color={ICON_COLOR}
    style={{ textAlign: 'center' }}
  />
