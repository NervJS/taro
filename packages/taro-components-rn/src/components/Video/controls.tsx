// import React, { Component } from 'react'
// import View from '../View'
// import { ControlsProps } from './PropsType'
// import { formatTime } from './utils'
// import Styles from './style'
// import Image from '../Image'
// /**
//  * @typedef {Object} ControlsProps
//  * @property {Boolean} controls={controls}
//  * @property {Number} currentTime={this.currentTime}
//  * @property {Number} duration={this.state.duration}
//  * @property {Boolean} isPlaying={this.state.isPlaying}
//  * @property {Function} pauseFunc={this.pause}
//  * @property {Function} playFunc={this.play}
//  * @property {Function} seekFunc={this.seek}
//  * @property {Boolean} showPlayBtn={showPlayBtn}
//  * @property {Boolean} showProgress={showProgress}
//  */
// class Controls extends Component<ControlsProps, any> {
//   visible = false
//   isDraggingProgressBall = false

//   /** @type {number} */
//   hideControlsTimer

//   progressDimentions = {
//     left: 0,
//     right: 0,
//     width: 0
//   }

//   calcPercentage = pageX => {
//     let pos = pageX - this.progressDimentions.left
//     pos = Math.max(pos, 0)
//     pos = Math.min(pos, this.progressDimentions.width)
//     return pos / this.progressDimentions.width
//   }

//   getControlsRef = ref => {
//     if (!ref) return
//     this.controlsRef = ref
//   }
//   getCurrentTimeRef = ref => {
//     if (!ref) return
//     this.currentTimeRef = ref
//   }
//   getProgressBallRef = ref => {
//     if (!ref) return
//     this.progressBallRef = ref
//   }

//   setCurrentTime (time) {
//     if (this.currentTimeRef) {
//       this.currentTimeRef.innerHTML = formatTime(time)
//     }
//   }
//   setProgressBall (percentage) {
//     if (this.progressBallRef) {
//       this.progressBallRef.style.left = `${percentage * 100}%`
//     }
//   }

//   toggleVisibility (nextVisible) {
//     const visible = nextVisible === undefined ? !this.visible : nextVisible
//     if (visible) {
//       this.hideControlsTimer && clearTimeout(this.hideControlsTimer)
//       if (this.props.isPlaying) {
//         this.hideControlsTimer = setTimeout(() => {
//           this.toggleVisibility(false)
//         }, 2000)
//       }
//       if (this.controlsRef) {
//         this.controlsRef.style.visibility = 'visible'
//       }
//     } else {
//       if (this.controlsRef) {
//         this.controlsRef.style.visibility = 'hidden'
//       }
//     }
//     this.visible = !!visible
//   }

//   onDragProgressBallStart = () => {
//     this.isDraggingProgressBall = true
//     this.hideControlsTimer && clearTimeout(this.hideControlsTimer)
//   }
//   onClickProgress = e => {
//     e.stopPropagation()
//     const seekFunc = this.props.seekFunc
//     const percentage = this.calcPercentage(e.pageX)
//     seekFunc(percentage * this.props.duration)
//     this.toggleVisibility(true)
//   }
//   bindTouchEvents = () => {
//     let percentage = 0
//     const touchMove = e => {
//       if (!this.isDraggingProgressBall) return
//       const touchX = e.touches[0].pageX
//       percentage = this.calcPercentage(touchX)
//       this.setProgressBall(percentage)
//     }
//     const touchEnd = e => {
//       if (!this.isDraggingProgressBall) return
//       const seekFunc = this.props.seekFunc
//       this.isDraggingProgressBall = false
//       seekFunc(percentage * this.props.duration)
//       this.toggleVisibility(true)
//     }

//     // document.body.addEventListener('touchmove', touchMove)
//     // document.body.addEventListener('touchend', touchEnd)
//     // document.body.addEventListener('touchcancel', touchEnd)
//     // return () => {
//     //   document.body.removeEventListener('touchmove', touchMove)
//     //   document.body.removeEventListener('touchend', touchEnd)
//     //   document.body.removeEventListener('touchcancel', touchEnd)
//     // }
//   }

//   componentDidMount () {
//     // this.unbindTouchEvents = this.bindTouchEvents()
//   }
//   componentWillUnmount () {
//     // this.unbindTouchEvents()
//   }

//   render () {
//     const { controls, currentTime, duration, isPlaying, pauseFunc, playFunc, showPlayBtn, showProgress } = this.props
//     const formattedDuration = formatTime(duration as number)
//     let playBtn

//     if (!showPlayBtn) {
//       return null
//     } else if (isPlaying) {
//       playBtn = <Image src={require('../../assets/video/play.png')} style={[Styles['taro-video-control-button'], Styles['taro-video-control-button-pause']]} onClick={pauseFunc} />;
//     } else {
//       playBtn = <Image src={require('../../assets/video/play.png')} style={[Styles['taro-video-control-button'], Styles['taro-video-control-button-pause']]} onClick={pauseFunc} />;
//     }

//     return (
//       <View style={[Styles['taro-video-bar'], Styles['taro-video-bar-full']]} ref={this.getControlsRef}>
//         {controls && (
//           <View style={Styles['taro-video-controls']}>
//             {playBtn}
//             {showProgress && (
//               <View style={Styles['taro-video-current-time']} ref={this.getCurrentTimeRef}>
//                 {formatTime(currentTime as number)}
//               </View>
//             )}
//             {showProgress && (
//               <View style={Styles['taro-video-progress-container']} onClick={this.onClickProgress}>
//                 <View
//                   style={Styles['taro-video-progress']}
//                   ref={ref => {
//                     // if (ref !== null) {
//                     //   const rect = ref.getBoundingClientRect()
//                     //   this.progressDimentions.left = rect.left
//                     //   this.progressDimentions.right = rect.right
//                     //   this.progressDimentions.width = rect.width
//                     // }
//                   }}>
//                   <View style={Styles['taro-video-progress-buffered']} style={{ width: '100%' }} />
//                   <View style={Styles['taro-video-ball']} ref={this.getProgressBallRef} onTouchStart={this.onDragProgressBallStart} style={`left: ${formattedDuration ? (this.currentTime / duration) * 100 : 0}%`}>
//                     <View style={Styles['taro-video-inner']} />
//                   </View>
//                 </View>
//               </View>
//             )}
//             {showProgress && <View style={Styles['taro-video-duration']}>{formattedDuration}</View>}
//           </View>
//         )}
//         {this.props.children}
//       </View>
//     )
//   }
// }

// export default Controls
