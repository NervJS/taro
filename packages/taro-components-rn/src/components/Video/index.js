/**
 * ✔ src
 * ✔ initialTime
 * ✘ duration
 * ✔ controls
 * ✘ danmuList
 * ✘ danmuBtn
 * ✘ enableDanmu
 * ✔ autoplay
 * ✔ loop
 * ✔ muted
 * ✘ pageGesture
 * ✘ direction
 * ✔ showProgress
 * ✔ showFullscreenBtn
 * ✔ showPlayBtn
 * ✔ showCenterPlayBtn
 * ✘ enableProgressGesture
 * ✔ objectFit
 * ✔ poster
 * ✔ onPlay
 * ✔ onPause
 * ✔ onEnded
 * ✔ onTimeupdate
 * ✔ onFullscreenchange
 * ✔ onWaiting
 * ✔ onError
 * ✘ onProgress
 *
 * @flow
 */

import * as React from 'react'
import { Dimensions } from 'react-native'
import { Video, ScreenOrientation } from 'expo'
import Player from './player'
import utils from '../../utils'

const resizeModeMap = {
  contain: Video.RESIZE_MODE_CONTAIN,
  fill: Video.RESIZE_MODE_STRETCH,
  cover: Video.RESIZE_MODE_COVER
}

type Props = {
  src: string,
  initialTime?: number,
  controls?: boolean,
  autoplay?: boolean,
  loop?: boolean,
  muted?: boolean,
  showProgress?: boolean,
  showFullscreenBtn?: boolean,
  showPlayBtn?: boolean,
  showCenterPlayBtn?: boolean,
  objectFit?: 'contain' | 'fill' | 'cover',
  poster?: string,
  onPlay?: Function,
  onPause?: Function,
  onEnded?: Function,
  onTimeupdate?: Function,
  onFullscreenchange?: Function,
  onWaiting?: Function,
  onError?: Function
}

type State = {
  isPortrait: boolean
}

class _Video extends React.Component<Props, State> {
  props: Props
  state: State = {
    isPortrait: true
  }

  static defaultProps = {
    initialTime: 0,
    controls: true,
    showProgress: true,
    showFullscreenBtn: true,
    showPlayBtn: true,
    objectFit: 'contain',
    onPlay: utils.noop,
    onPause: utils.noop,
    onEnded: utils.noop,
    onTimeupdate: utils.noop,
    onFullscreenchange: utils.noop,
    onWaiting: utils.noop,
    onError: utils.noop
  }

  _onPlaybackStatusUpdate = (playbackStatus) => {
    if (!playbackStatus.isLoaded) {
      if (playbackStatus.error) {
        this.props.onError(playbackStatus.error)
      }
      return
    }

    this.props.onTimeupdate(playbackStatus.positionMillis)

    if (playbackStatus.isPlaying) {
      this.props.onPlay()
    } else {
      this.props.onPause()
    }

    if (playbackStatus.isBuffering) {
      this.props.onWaiting()
    }

    if (playbackStatus.didJustFinish && !playbackStatus.isLooping) {
      this.props.onEnded()
    }
  }

  // _handleVideoRef = (component) => {
  //   const playbackObject = component
  //   playbackObject.loadAsync({ uri: this.props.src }, {
  //     positionMillis: this.props.initialTime,
  //     shouldPlay: this.props.autoplay,
  //     isLooping: this.props.loop,
  //     isMuted: this.props.isMuted
  //   })
  //   playbackObject.setOnPlaybackStatusUpdate(this._onPlaybackStatusUpdate)
  // }

  componentDidMount () {
    ScreenOrientation.allow(ScreenOrientation.Orientation.ALL)
    Dimensions.addEventListener(
      'change',
      this.orientationChangeHandler.bind(this)
    )
  }

  componentWillUnmount() {
    ScreenOrientation.allow(ScreenOrientation.Orientation.PORTRAIT)
    Dimensions.removeEventListener('change', this.orientationChangeHandler)
  }

  orientationChangeHandler(dims) {
    const { width, height } = dims.window
    const isLandscape = width > height
    this.setState({ isPortrait: !isLandscape })
    this.props.onFullscreenchange({ detail: { fullScreen: !!isLandscape, direction: isLandscape ? 'horizontal' : 'vertical' } })
    ScreenOrientation.allow(ScreenOrientation.Orientation.ALL)
  }

  switchToLandscape = () => {
    ScreenOrientation.allow(ScreenOrientation.Orientation.LANDSCAPE)
  }

  switchToPortrait = () => {
    ScreenOrientation.allow(ScreenOrientation.Orientation.PORTRAIT)
  }

  render () {
    return (
      <Player
        videoProps={{
          source: { uri: 'http://d23dyxeqlo5psv.cloudfront.net/big_buck_bunny.mp4' },
          resizeMode: resizeModeMap[this.props.objectFit],
          posterSource: { uri: this.props.poster },
          positionMillis: this.props.initialTime,
          shouldPlay: this.props.autoplay,
          isLooping: this.props.loop,
          isMuted: this.props.isMuted
        }}
        controls={this.props.controls}
        showProgress={this.props.showProgress}
        showFullscreenBtn={this.props.showFullscreenBtn}
        showPlayBtn={this.props.showPlayBtn}
        showCenterPlayBtn={this.props.showCenterPlayBtn}
        playbackCallback={this._onPlaybackStatusUpdate}
        isPortrait={this.state.isPortrait}
        switchToLandscape={this.switchToLandscape}
        switchToPortrait={this.switchToPortrait}
      />
    )
  }
}

export default _Video
