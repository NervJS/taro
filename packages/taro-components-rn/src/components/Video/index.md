## API

| 是否支持 | 属性           | 类型        | 默认值 | 说明                                                         |
| -------- | -------------- | ----------- | ------ | ------------------------------------------------------------ |
|    | autoPauseIfNavigate | boolean | true |当跳转到其它小程序页面时，是否自动暂停本页面的视频
|    | autoPauseIfOpenNative | boolean | true |当跳转到其它微信原生页面时，是否自动暂停本页面的视频
|    | direction | number |  |设置全屏时视频的方向，不指定则根据宽高比自动判断
|    | title | string |  |视频的标题，全屏时在顶部展示
| √ | autoplay | boolean | false |是否自动播放
| √ | controls | boolean | true |是否显示默认播放控件（播放/暂停按钮、播放进度、时间）
| √ | danmuBtn | boolean | false |是否显示弹幕按钮，只在初始化时有效，不能动态变更
| √ | danmuList | Array.<Danmu> | [] |弹幕列表
| √ | duration | number |  |指定视频时长
| √ | enableDanmu | boolean | false |是否展示弹幕，只在初始化时有效，不能动态变更
| √ | enablePlayGesture | boolean | false |是否开启播放手势，即双击切换播放/暂停
| √ | enableProgressGesture | boolean | true |是否开启控制进度的手势
| √ | initialTime | number | 0 |指定视频初始播放位置
| √ | loop | boolean | false |是否循环播放
| √ | muted | boolean | false |是否静音播放
| √ | objectFit | string | contain |当视频大小与 video 容器大小不一致时，视频的表现形式
| √ | onEnded | Function |  |当播放到末尾时触发 ended 事件
| √ | onError | Function |  |视频播放出错时触发
| √ | onFullscreenChange | Function |  |视频进入和退出全屏时触发，event.detail = {fullScreen, direction}，direction 有效值为 vertical 或 horizontal
| √ | onPause | Function |  |当暂停播放时触发 pause 事件
| √ | onPlay | Function |  |当开始/继续播放时触发play事件
| √ | onProgress | Function |  |加载进度变化时触发，只支持一段加载。event.detail = {buffered}，百分比
| √ | onTimeUpdate | Function |  |播放进度变化时触发，event.detail = {currentTime, duration}。触发频率 250ms 一次
| √ | onWaiting | Function |  |视频出现缓冲时触发
| √ | playBtnPosition | string | bottom |播放按钮的位置
| √ | poster  | string |  |视频封面的图片网络资源地址。若 controls 属性值为 false 则设置 poster 无效
| √ | showCenterPlayBtn | boolean | true |是否显示视频中间的播放按钮
| √ | showFullscreenBtn | boolean | true |是否显示全屏按钮
| √ | showMuteBtn | boolean | false |是否显示静音按钮
| √ | showPlayBtn | boolean | true |是否显示视频底部控制栏的播放按钮
| √ | showProgress | boolean | true |若不设置，宽度大于240时才会显示
| √ | src | string |  |要播放视频的资源地址
| √ | vslideGesture | boolean | false |在非全屏模式下，是否开启亮度与音量调节手势（同 pageGesture）
| √ | vslideGestureInFullscreen | boolean | true |在全屏模式下，是否开启亮度与音量调节手势
