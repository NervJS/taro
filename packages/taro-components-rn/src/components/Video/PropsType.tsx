import { VideoProps } from '@tarojs/components/types/Video'
import * as React from 'react'
import { StyleProp, ViewStyle } from 'react-native'

export interface ViewProps {
  style?: StyleProp<ViewStyle>;
  children?: React.ReactNode;
}

export interface onFullscreenChangeEventDetail extends VideoProps.onFullscreenChangeEventDetail {
  fullscreenUpdate: 0 | 1 | 2 | 3;
}

export interface ControlsProps {
  controls?: boolean;
  currentTime?: number;
  duration?: number;
  isPlaying?: boolean;
  pauseFunc?: () => void;
  playFunc?: () => void;
  seekFunc?: () => void;
  showPlayBtn?: boolean;
  showProgress?: boolean;
}
