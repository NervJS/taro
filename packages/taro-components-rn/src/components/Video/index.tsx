/**
 * Video组件的样式参考了[uni-app](https://github.com/dcloudio/uni-app/tree/master/packages/uni-h5)的实现
 */

import React, { Component } from 'react'
// import Danmu from './danmu'
// import Controls from './controls'
import { formatTime } from './utils'
import { VideoProps } from '@tarojs/components/types/Video'
import {
  AVPlaybackStatus,
  Video,
  VideoFullscreenUpdateEvent,
  VideoReadyForDisplayEvent,
} from 'expo-av'
import Styles from './style'
import View from '../View'
import Text from '../Text'
import Image from '../Image'
import { onFullscreenChangeEventDetail } from './PropsType'
/**
 * @typedef {Object} Danmu
 * @property {string} text 弹幕文字
 * @property {string} color 弹幕颜色
 * @property {number} [time] 弹幕时间
 */

/**
 * Video组件参数
 * @typedef {Object} VideoProps
 * @property {string} src 要播放视频的资源地址
 * @property {boolean} [autoPauseIfNavigate=true] 当跳转到其它小程序页面时，是否自动暂停本页面的视频
 * @property {boolean} [autoPauseIfOpenNative=true] 当跳转到其它微信原生页面时，是否自动暂停本页面的视频
 * @property {boolean} [autoplay=false] 是否自动播放
 * @property {boolean} [controls=true] 是否显示默认播放控件（播放/暂停按钮、播放进度、时间）
 * @property {boolean} [danmuBtn=false] 是否显示弹幕按钮，只在初始化时有效，不能动态变更
 * @property {Array.<Danmu>} [danmuList=[]] 弹幕列表
 * @property {boolean} [enableDanmu=false] 是否展示弹幕，只在初始化时有效，不能动态变更
 * @property {boolean} [enablePlayGesture=false] 是否开启播放手势，即双击切换播放/暂停
 * @property {boolean} [enableProgressGesture=true] 是否开启控制进度的手势
 * @property {number} [initialTime=0] 指定视频初始播放位置
 * @property {boolean} [loop=false] 是否循环播放
 * @property {boolean} [muted=false] 是否静音播放
 * @property {string} [objectFit=contain] 当视频大小与 video 容器大小不一致时，视频的表现形式
 * @property {string} [playBtnPosition=bottom] 播放按钮的位置
 * @property {boolean} [showCenterPlayBtn=true] 是否显示视频中间的播放按钮
 * @property {boolean} [showFullscreenBtn=true] 是否显示全屏按钮
 * @property {boolean} [showMuteBtn=false] 是否显示静音按钮
 * @property {boolean} [showPlayBtn=true] 是否显示视频底部控制栏的播放按钮
 * @property {boolean} [showProgress=true] 若不设置，宽度大于240时才会显示
 * @property {boolean} [vslideGesture=false] 在非全屏模式下，是否开启亮度与音量调节手势（同 pageGesture）
 * @property {boolean} [vslideGestureInFullscreen=true] 在全屏模式下，是否开启亮度与音量调节手势
 * @property {number} [direction]  设置全屏时视频的方向，不指定则根据宽高比自动判断
 * @property {number} [duration]  指定视频时长
 * @property {string} [poster]  视频封面的图片网络资源地址。若 controls 属性值为 false 则设置 poster 无效
 * @property {string} [title]  视频的标题，全屏时在顶部展示
 * @property {Function} [onPlay] 当开始/继续播放时触发play事件
 * @property {Function} [onPause] 当暂停播放时触发 pause 事件
 * @property {Function} [onEnded] 当播放到末尾时触发 ended 事件
 * @property {Function} [onTimeUpdate] 播放进度变化时触发，event.detail = {currentTime, duration} 。触发频率 250ms 一次
 * @property {Function} [onFullscreenChange] 视频进入和退出全屏时触发，event.detail = {fullScreen, direction}，direction 有效值为 vertical 或 horizontal
 * @property {Function} [onWaiting] 视频出现缓冲时触发
 * @property {Function} [onError] 视频播放出错时触发
 * @property {Function} [onProgress] 加载进度变化时触发，只支持一段加载。event.detail = {buffered}，百分比
 */

const ObjectFit = {
  contain: Video.RESIZE_MODE_CONTAIN,
  fill: Video.RESIZE_MODE_STRETCH,
  cover: Video.RESIZE_MODE_COVER,
}

declare const global: any

global._taroVideoMap = {}

interface Props extends VideoProps {
  onLoad: () => void;
}

class _Video extends Component<Props, any> {
  /** @type {VideoProps} */
  static defaultProps: Props = {
    id: '',
    src: '',
    autoPauseIfNavigate: true,
    autoPauseIfOpenNative: true,
    autoplay: false,
    controls: true,
    danmuBtn: false,
    danmuList: [],
    enableDanmu: false,
    enablePlayGesture: false,
    enableProgressGesture: true,
    initialTime: 0,
    loop: false,
    muted: false,
    objectFit: 'contain',
    playBtnPosition: 'bottom',
    showCenterPlayBtn: true,
    showFullscreenBtn: true,
    showMuteBtn: false,
    showPlayBtn: true,
    showProgress: true,
    vslideGesture: false,
    vslideGestureInFullscreen: true,
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onLoad: () => {},
  };

  /** @type {HTMLVideoElement} */
  videoRef: Video;

  /** @type {number} */
  currentTime = 0;

  progressDimentions = {
    left: 0,
    right: 0,
    width: 0,
  };

  getVideoRef: (ref: any) => void;
  isDraggingProgress: any;
  duration: any;
  toastVolumeRef: any;
  toastVolumeBarRef: any;
  toastProgressTitleRef: any;
  toastProgressRef: any;
  getControlsRef: (ref: any) => void;
  getDanmuRef: (ref: any) => void;
  getToastProgressRef: (ref: any) => void;
  getToastProgressTitleRef: (ref: any) => void;
  getToastVolumeRef: (ref: any) => void;
  getToastVolumeBarRef: (ref: any) => void;
  unbindTouchEvents: () => void;

  constructor({ props, context }: { props: Props; context: any }) {
    super(props, context)
    const stateObj = this.props
    this.videoRef = (React.createRef() as unknown) as Video
    this.state = Object.assign(
      {
        duration: null,
        isPlaying: false,
        isFirst: true,
        enableDanmu: false,
        isFullScreen: false,
        isMute: false,
      },
      stateObj
    )
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  onEnded = (e: any): void => {
    this.setState({
      // isFirst: true,
      isEnded: true,
    })
    if (!this.props.loop) this.pause()
    this.props.onEnded && this.props.onEnded(e)
  };

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  onPlay = (e: any): void => {
    this.props.onPlay && this.props.onPlay(e)
    if (!this.state.isPlaying) {
      this.play()
    }
  };

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  onPause = (e: any): void => {
    this.props.onPause && this.props.onPause(e)
    if (this.state.isPlaying) {
      this.setState({
        isPlaying: false,
      })
    }
  };

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  onError = (e: string): void => {
    if (this.props.onError) {
      const error: any = Object.defineProperty({}, 'detail', {
        enumerable: true,
        value: { errMsg: e },
      })
      // @ts-ignore
      this.props.onError(error)
    }
  };

  onLoad = (status: AVPlaybackStatus): void => {
    const { durationMillis = 0 }: any = status
    this.setState({
      duration: durationMillis,
    })
    this.props.onLoad && this.props.onLoad()
  };

  clickPlayBtn = ():void => {
    const { isEnded } = this.state
    isEnded && this.seek(0)
    this.videoRef && this.videoRef.playAsync()
    this.setState({
      shouldPlay: true,
      isFirst: false,
    })
  }

  play = (): void => {
    const { isEnded } = this.state
    isEnded && this.seek(0)
    this.videoRef.playAsync()
    this.setState({
      shouldPlay: true,
      isPlaying: true,
      isFirst: false,
    })
  };

  pause = (): void => {
    this.setState({
      isPlaying: false,
      shouldPlay: false,
    })
  };

  stop = (): void => {
    this.pause()
    this.seek(0)
    this.setState({
      isPlaying: false,
    })
  };

  seek = (position: number): void => {
    this.videoRef.setStatusAsync({
      positionMillis: position,
    })
  };

  showStatusBar = (): void => {
    console.error('暂不支持 videoContext.showStatusBar')
  };

  hideStatusBar = (): void => {
    console.error('暂不支持 videoContext.hideStatusBar')
  };

  requestFullScreen = (): void => {
    // @ts-ignore
    this.videoRef.presentFullscreenPlayer()
  };

  exitFullScreen = (): void => {
    // @ts-ignore
    this.videoRef.dismissFullscreenPlayer()
  };

  componentDidMount(): void {
    const getRef = (refName: string) => {
      const { id } = this.props
      return (ref: any) => {
        if (!ref) return
        // @ts-ignore
        this[refName] = ref
        // @ts-ignore
        id && (global._taroVideoMap[id] = ref)
      }
    }
    this.getVideoRef = getRef('videoRef')
  }

  static getDerivedStateFromProps(nProps: VideoProps): VideoProps {
    return nProps
  }

  onLoadedMetaData = (event: VideoReadyForDisplayEvent): void => {
    const { naturalSize, status } = event
    // @ts-ignore
    status.duration = status.durationMillis
    // @ts-ignore
    this.props.onLoadedMetaData && this.props.onLoadedMetaData({ detail: { ...naturalSize, ...status } })
  };

  onFullscreenChange = (event: VideoFullscreenUpdateEvent): void => {
    const { fullscreenUpdate, status } = event
    const fullScreen: boolean = fullscreenUpdate === Video.FULLSCREEN_UPDATE_PLAYER_WILL_PRESENT || fullscreenUpdate === Video.FULLSCREEN_UPDATE_PLAYER_DID_PRESENT
    const detail: onFullscreenChangeEventDetail = {
      fullScreen: fullscreenUpdate === Video.FULLSCREEN_UPDATE_PLAYER_WILL_PRESENT || fullscreenUpdate === Video.FULLSCREEN_UPDATE_PLAYER_DID_PRESENT,
      fullscreenUpdate,
      direction: 1,
      ...status,
    }
    if (this.state.isFullScreen !== fullScreen) {
      this.setState({
        isFullScreen: fullScreen,
      }, () => {
        // @ts-ignore
        this.props.onFullscreenChange && this.props.onFullscreenChange({ detail })
      })
    }
  };

  onPlaybackStatusUpdate = (event: AVPlaybackStatus): void => {
    // @ts-ignore
    this.props.onTimeUpdate && this.props.onTimeUpdate(event)
    // @ts-ignore
    const { didJustFinish, isPlaying } = event
    if (didJustFinish) {
      this.onEnded(event)
    }

    if (isPlaying !== this.state.isPlaying) {
      this.setState(
        {
          isPlaying,
          isFirst: isPlaying ? false : this.state.isFirst
        },
        () => {
          isPlaying && this.onPlay(event)
          !isPlaying && !this.state.isFirst && this.onPause(event)
        }
      )
    }
  };

  render(): JSX.Element {
    const {
      src,
      autoplay,
      style,
      initialTime,
      loop,
      muted,
      objectFit = 'contain',
      poster,
      controls,
      showCenterPlayBtn,
    } = this.props
    const { isFullScreen, shouldPlay } = this.state
    const duration = formatTime(
      this.props.duration || this.state.duration || null
    )
    const videoProps = {
      source: { uri: src },
      shouldPlay: shouldPlay || autoplay,
      posterSource: controls ? { uri: poster } : undefined,
      usePoster: !!controls,
      isLooping: loop,
      isMuted: muted,
      positionMillis: initialTime,
      style: Object.assign({ width: '100%', height: '100%' }, style as Record<string, unknown>),
      ref: this.getVideoRef,
      resizeMode: ObjectFit[objectFit],
      useNativeControls: controls,
      onError: this.onError,
      onLoad: this.onLoad,
      onReadyForDisplay: this.onLoadedMetaData,
      onFullscreenUpdate: this.onFullscreenChange,
      onPlaybackStatusUpdate: this.onPlaybackStatusUpdate,
    }

    // 第一次不显示又无法自动播放
    const showPlayBtn = this.state.isFirst ? true : showCenterPlayBtn

    const videoNode = (
      <View
        style={[
          Styles['taro-video-container'],
          isFullScreen && Styles['taro-video-type-fullscreen'],
        ]}
      >
        <Video {...videoProps} />
        {showPlayBtn && !this.state.isPlaying && (
          <View style={Styles['taro-video-cover']}>
            <Image
              src={require('../../assets/video/play.png')}
              style={Styles['taro-video-cover-play-button']}
              onClick={this.clickPlayBtn}
            />
            <Text style={Styles['taro-video-cover-duration']}>{duration}</Text>
          </View>
        )}
        {this.props.children}
      </View>
    )
    return (
      <View style={[Styles['taro-video'], this.props.style as Record<string, unknown>]}>
        {videoNode}
      </View>
    )
  }
}

export default _Video
