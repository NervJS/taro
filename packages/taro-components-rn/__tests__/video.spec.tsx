// eslint-disable-next-line no-use-before-define
import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Video from '../src/components/Video'

describe('Video', () => {
  const videoOpts = {
    src: 'http://wxsnsdy.tc.qq.com/105/20210/snsdyvideodownload?filekey=30280201010421301f0201690402534804102ca905ce620b1241b726bc41dcff44e00204012882540400&bizid=1023&hy=SH&fileparam=302c020101042530230204136ffd93020457e3c4ff02024ef202031e8d7f02030f42400204045a320a0201000400',
    controls: true,
    autoplay: false,
    poster: 'http://misc.aotu.io/booxood/mobile-video/cover_900x500.jpg',
    initialTime: 30,
    id: 'video',
    loop: false,
    muted: false
  }

  it('render Video', () => {
    const wrapper = shallow(<Video {...videoOpts} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
