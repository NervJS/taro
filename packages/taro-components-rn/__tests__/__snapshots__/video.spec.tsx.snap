// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video render Video 1`] = `
<_Clickable
  hoverStartTime={20}
  hoverStayTime={70}
  style={
    Array [
      Object {
        "height": 225,
        "lineHeight": 0,
        "overflow": "hidden",
        "position": "relative",
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <_Clickable
    hoverStartTime={20}
    hoverStayTime={70}
    style={
      Array [
        Object {
          "backgroundColor": "#000",
          "bottom": 0,
          "height": "100%",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        },
        false,
      ]
    }
  >
    <Video
      isLooping={false}
      isMuted={false}
      onError={[Function]}
      onFullscreenUpdate={[Function]}
      onLoad={[Function]}
      onPlaybackStatusUpdate={[Function]}
      onReadyForDisplay={[Function]}
      positionMillis={30}
      posterSource={
        Object {
          "uri": "http://misc.aotu.io/booxood/mobile-video/cover_900x500.jpg",
        }
      }
      resizeMode="contain"
      shouldPlay={false}
      source={
        Object {
          "uri": "http://wxsnsdy.tc.qq.com/105/20210/snsdyvideodownload?filekey=30280201010421301f0201690402534804102ca905ce620b1241b726bc41dcff44e00204012882540400&bizid=1023&hy=SH&fileparam=302c020101042530230204136ffd93020457e3c4ff02024ef202031e8d7f02030f42400204045a320a0201000400",
        }
      }
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
      useNativeControls={true}
      usePoster={true}
    />
    <_Clickable
      hoverStartTime={20}
      hoverStayTime={70}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(1, 1, 1, 0.5)",
          "bottom": 0,
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
          "zIndex": 1,
        }
      }
    >
      <_Clickable
        hoverStartTime={20}
        hoverStayTime={70}
        onClick={[Function]}
        src={[Function]}
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
      <_Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 16,
            "marginTop": 10,
          }
        }
      />
    </_Clickable>
  </_Clickable>
</_Clickable>
`;
